src/resolver.js

```diff
@@ -0,0 +1,162 @@
+const REGISTRY_ENTRY_REGEXP = /^skyns:\/\/(?<publickey>[a-zA-Z0-9%]+)\/(?<datakey>[a-zA-Z0-9%]+)$/;
+
+class ResolutionError extends Error {}
+
+/**
+ * @typedef {Object} HSRecord
+ * This documentation only mentiones properties used by this service. If you need more information,
+ * visit https://hsd-dev.org/guides/resource-records.html.
+ *
+ * @property {string}   type  Record type. In case of this application, we only care about TXT records.
+ * @property {string[]} [txt] Values for TXT records.
+ *                            There are many other possible properties besides .txt, but within this
+ *                            application we only care about TXT records.
+ *
+ * @typedef {Object} RegistryEntry
+ * @property {string} publickey
+ * @property {string} datakey
+ *
+ * @typedef {Object} ResolvedSkylink
+ * @property {string} skylink
+ *
+ * @typedef {Object} ResolvedRegistryEntry
+ * @property {RegistryEntry} registry
+ *
+ * @typedef {Object} HSClientOptions
+ * @property {string} network Target Handshake network
+ * @property {string} host    Hostname of the Handshake node server
+ * @property {string} port    Port of the Handshake node server
+ * @property {string} apiKey  API key to be used to connect to the Handshake server
+ */
+
+/**
+ * Resolver class is responsible for validating the requested domain names
+ * and decoding TXT records for them, as long as they exist.
+ */
+class Resolver {
+  /**
+   * Create a Resolver instance.
+   *
+   * @param {HSClientOptions} hsClientOptions
+   */
+  constructor(hsClientOptions) {
+    this.hsClient = new NodeClient(hsClientOptions);
+    this.cache = new NodeCache({ stdTTL: 300 }); // cache for 5 minutes
+  }
+  /**
+   * Try to resolve given Handshake domain into a skylink or skynet registry entry.
+   *
+   * @param {string} rawDomainName Domain name
+   * @throws {ResolutionError}
+   * @returns {ResolvedSkylink | ResolvedRegistryEntry}
+   */
+  async resolve(rawDomainName) {
+    const domain = punycode.toASCII(rawDomainName);
+    const records = await this.getDomainRecords(domain);
+    if (!records) {
+      throw new ResolutionError(`No records found for ${domain}`);
+    }
+
+    const record = this.findSkynetCompatibleRecord(records);
+    if (!record) {
+      throw new Error(`No skynet compatible records found in dns records of ${domain}`);
+    }
+
+    return this.createResponseFromCompatibleRecord(record);
+  }
+
+  /**
+   * Get Handshake domain records for a given domain
+   *
+   * @param {string} domainName Domain name to be checked for DNS link records.
+   * @returns {Promise<HSRecord[]>}
+   */
+  async getDomainRecords(name) {
+    if (this.cache.has(name)) {
+      return this.cache.get(name);
+    }
+
+    const response = await this.hsClient.execute("getnameresource", [name]);
+    const records = response?.records ?? null;
+
+    logger.info(`${name} => ${JSON.stringify(records)}`);
+
+    this.cache.set(name, records);
```

Comment:
nit Q: Should we set the cache value if the records are `null`? We might want to check the response error in order to determine whether it's a permanent situation or a temporary outage. We should also log that error, in case it's permanent.

Reply:
I'd consider it as a follow up. I tried not to change any of the existing functionalities -- just restructure the code to be able to test it.

Reply:
F/U is fine.

Reply:
Created [SKY-1360](https://linear.app/skynetlabs/issue/SKY-1360/consider-not-caching-empty-responses-from-handshake-node).